File: Ordenamiento/QuickSort.java
package Ordenamiento;

import java.util.Arrays;
import java.util.Stack;

public class QuickSort {
    public static void quickSort(int[] array) {
        int low = 0;
        int high = array.length - 1;

        // Crear una pila para almacenar los límites del subarray
        Stack<Integer> stack = new Stack<>();
        stack.push(low);
        stack.push(high);

        // Iterativamente procesar subarrays
        while (!stack.isEmpty()) {
            high = stack.pop();
            low = stack.pop();

            // Obtener el índice del pivote después de particionar
            int pivotIndex = partition(array, low, high);

            // Si hay elementos a la izquierda del pivote, agregar el subarray izquierdo
            if (pivotIndex - 1 > low) {
                stack.push(low);
                stack.push(pivotIndex - 1);
            }

            // Si hay elementos a la derecha del pivote, agregar el subarray derecho
            if (pivotIndex + 1 < high) {
                stack.push(pivotIndex + 1);
                stack.push(high);
            }
        }
    }

    // Función para particionar el array y devolver el índice del pivote
    static int partition(int[] array, int low, int high) {
        int pivot = array[high];
        int i = low - 1;

        for (int j = low; j < high; j++) {
            if (array[j] <= pivot) {
                i++;
                // Intercambiar array[i] y array[j]
                int temp = array[i];
                array[i] = array[j];
                array[j] = temp;
            }
        }

        // Intercambiar array[i+1] y array[high] (pivote)
        int temp = array[i + 1];
        array[i + 1] = array[high];
        array[high] = temp;

        return i + 1;
    }

    public static void main(String[] args) {
        int[] array = {5, 2, 9, 1, 5, 6};

        System.out.println("Array antes del ordenamiento: " + Arrays.toString(array));

        quickSort(array);

        System.out.println("Array después del ordenamiento: " + Arrays.toString(array));
    }
}

File: Ordenamiento/QuickSort.py
import cProfile
import random
import time

def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def quick_sort(arr, ascending=True):
    stack = [(0, len(arr) - 1)]
    while stack:
        low, high = stack.pop()
        if low < high:
            pi = partition(arr, low, high)
            stack.append((low, pi - 1))
            stack.append((pi + 1, high))

    if not ascending:
        arr.reverse()

def profile_quick_sort(arr, ascending=True):
    profiler = cProfile.Profile()
    profiler.enable()
    quick_sort(arr, ascending)
    profiler.disable()
    profiler.print_stats(sort='cumulative')

def sort_and_time(arr, ascending=True):
    start_time = time.time()
    quick_sort(arr, ascending)
    end_time = time.time()
    return end_time - start_time

if __name__ == "__main__":
    size = int(input("Ingrese la cantidad de elementos: "))
    array = [random.randint(1, 100) for _ in range(size)]
    print("Array antes del ordenamiento:", array)

    choice = input("¿Desea ordenar de manera ascendente (Sí/No)? ").lower()
    ascending_order = choice.startswith('s')

    profile_choice = input("¿Desea perfilar el código (Sí/No)? ").lower()
    profiling_enabled = profile_choice.startswith('s')

    if profiling_enabled:
        profile_quick_sort(array, ascending_order)
    else:
        quick_sort(array, ascending_order)

    print("Array después del ordenamiento:", array)

    sort_again = input("¿Desea ordenar el array nuevamente (Sí/No)? ").lower()
    if sort_again.startswith('s'):
        # Medir el tiempo para ordenar una lista ya ordenada
        start_time = time.time()
        quick_sort(array, ascending_order)
        end_time = time.time()
        print("Array ordenado nuevamente:", array)
        print("Tiempo de ordenamiento para un array ya ordenado:", end_time - start_time)